### ISSUE TITLE
[discussion] Self-Sovereign Identity

### ISSUE THREAD

**Issue by makoConstruct** [bonfire-networks/bonfire-app#491]
(If this is more suited to some ActivityPub forum, I can take it there instead. I broach it here because I think standards are ultimately need to be steered by implementations, and I think this implementation has great taste, and also, some of the obstacles are UX issues.)

Why I need SSI:

- It's hard to justify [using bonfire for contracts, transactions and accounting](https://github.com/bonfire-networks/bonfire_valueflows) or anything serious (any system that results in actions being taken in the real world) when instances could either impersonate their members to steal assets or, through negligence or as a result of an accident, go offline, lose their data, and permanently lock people out of their accounts.

- SSI can make the signup process frictionless. The user's browser can generate a key locally and they can sign up and authenticate using that, without having to go through an email registration process (the new account can later be given an email address, or merged into another account that has a recovery email or a recovery key registered). Friction is a huge deal. Any amount of friction makes a lot of legitimate social technologies infeasible. This might matter to bonfire if bonfire starts getting extensions that are so useful that they don't really need the rest of the social network for users to be interested in them, tools like when2meet or piratepad, where social network integration is useful, but not necessary, and where eliminating lowering barriers to participation is crucial.

So, this is important. But from what I understand, this is going to be difficult to do on ActivityPub, it seems like users are always referenced via their instance name? Which in turn answers to the DNS system?

**Comment by mayel** [comment]
Interesting topic indeed!

There's a few things on the backburner which I'd like to do in that direction:

1. Be able to enter a username with a different domain than one's instance (eg. if I (or a friend or family member) have registered the domain name mydomain.com and have signed up to the bonfire instance aroundthefire.net I should be able to identify as @mayel@mydomain.com rather than @mayel@aroundthefire.net) 
To do this, Bonfire could generate an appropriate webfinger JSON file that I'd have to upload to be served alongside my personal website at mydomain.com/.well-known/webfinger so that querying http://mydomain.com/.well-known/webfinger?resource=acct:@mayel@mydomain.com would return something like:
```
{
    "aliases": [
        "https://aroundthefire.net/@mayel",
        "https://aroundthefire.net/pub/actors/mayel"
    ],
    "links": [
        {
            "href": "https://aroundthefire.net/@maye",
            "rel": "http://webfinger.net/rel/profile-page",
            "type": "text/html"
        },
        {
            "href": "https://aroundthefire.net/pub/actors/mayel",
            "rel": "self",
            "type": "application/activity+json"
        },
        {
            "rel": "http://ostatus.org/schema/1.0/subscribe",
            "template": "https://aroundthefire.net/authorize_interaction?uri={uri}"
        }
    ],
    "subject": "acct:mayel@mydomain.com"
}
```
We'd have to make sure that instance migration (like Mastodon does) could still work, so that I can move from aroundthefire.net to burnitwithfire.xyz and retain followers while my mayel@mydomain.com username does not change.

2. Explore the possibility of optionally using just a domain, eg. @mydomain.com as a nick rather than requiring a domain + username. This would only work for people who have a personal domain name. 

The above are of course just stopgap workarounds that work with the way the fediverse works today, with it's dependency on the WebFinger standard and DNS based identity. Blockchain people would probably have many other suggestions but I'm not into that stuff...

**Comment by mayel** [comment]
There was also someone working on a distributed hash table project for the fediverse: https://socialhub.activitypub.rocks/c/software/hash2pub/48 for hashtags that could probably be adapted for something like this. But it seems to not have activity for a while.

**Comment by makoConstruct** [comment]
I spent some time searching and concepting for a way of doing the full thing, where users are always guaranteed to be able to recover control of their session key listing even if their primary host goes down, and... I think I've just ended up reinventing Holochain.

I'd bet that there simply isn't a way to do _SSI with reliable exit_ without having a general way of reaching consensus about the most recent version of the state of shared computations, which is, you know, a smart contract system. But I believe I can see a concept that does this in a fediversy way that wont require any gas fees.

**Comment by LiBa001** [comment]
Could [DIDs](https://www.w3.org/TR/did-1.0/) be used for that?
Like Bluesky does it with their AT protocol: https://atproto.com/guides/identity

I feel like this has probably been discussed before or I am missing something else.
I haven't looked too deep into these things (yet)

this also seems related to #1479 and #9

**Comment by mayel** [comment]
Yeah DIDs would be an obvious solution, of course the fediverse predates that standard so the question is could it be adopted as a progressive enhancement (not breaking or reducing interop with implementations who haven't implemented it)? I can only find this [basic FEP](https://codeberg.org/fediverse/fep/src/branch/main/fep/612d/fep-612d.md) along those lines so far.

Yeah those issues (plus the ones I've now linked to this one) are all related but simply aiming to improve UX a bit while working within the constraints of AP and WebFinger...

Re-reading the original issue though, this isn't really just about identity as in username and domain, but also about authentication as in private/public keys. And for that I'd really like to implement something like the [Secure Remote Password protocol](https://en.wikipedia.org/wiki/Secure_Remote_Password_protocol) like [Protonmail uses](https://proton.me/blog/encrypted-email-authentication) for authenticating on Bonfire but especially for deriving or encrypting the private key used for federation (currently for [HTTP signatures](https://swicg.github.io/activitypub-http-signature/), but could later be helpful for end-to-end-encrypted messages or other activities).
